tools/mason/MasonExternal.chpl

```diff
@@ -0,0 +1,431 @@
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+config const spackVersion = "releases/v0.11.2";
+
+use MasonUtils;
+use FileSystem;
+use MasonHelp;
+use MasonEnv;
+use Path;
+use TOML;
+
+proc masonExternal(args: [] string) {
+  try! {
+    if args.size < 3 {
+      masonExternalHelp();
+      exit(0);
+    }
+    else if args[2] == "-h" || args[2] == "--help" {
+      masonExternalHelp();
+      exit(0);
+    }
+    else if args[2] == "--setup" {
+      setupSpack();
+      exit(0);
+    }
+    if spackInstalled() {
+      select (args[2]) {
+        when 'search' do searchSpkgs(args);
+        when 'compiler' do compiler(args);
+        when 'install' do installSpkg(args);
+        when 'uninstall' do uninstallSpkg(args);
+        when 'info' do spkgInfo(args);
+        when 'find' do findSpkg(args);
+        when '--spec' do specHelp();
+        otherwise {
+          writeln('error: no such subcommand');
+          writeln('try mason external --help');
+          exit(1);
+        }
+      }
+    }
+  }
+  catch e: MasonError {
+    writeln(e.message());
+    exit(1);
+  }
+}
+
+/* Prints Spack spec help menu */
+private proc specHelp() {
+  const command = "spack help --spec";
+  const status = runSpackCommand(command);
+}
+
+private proc spackInstalled() throws {
+  if !isDir(MASON_HOME + "/spack") {
+    throw new MasonError("To use `mason external` call `mason external --setup`");
+  }
+  return true;
+}
+
+/* Spack installed to MASON_HOME/spack */
+proc setupSpack() throws {
+  writeln("Installing Spack backend ...");
+  const destination = MASON_HOME + "/spack/";
+  const clone = "git clone -q https://github.com/spack/spack " + destination;
+  const checkout = "git checkout -q " + spackVersion;
+  const status = runWithStatus(clone);
+  gitC(destination, checkout);
+  if status != 0 {
+    throw new MasonError("Spack installation failed");
+  }
+}
+
+/* Queries spack for package existance */
+proc spkgExists(spec: string) : bool {
+  const command = "spack list " + spec;
+  const status = runSpackCommand(command);
+  if status != 0 {
+    return false;
+  }
+  return true;
+}
+
+/* lists available spack packages */
+private proc listSpkgs() {
+  const command = "spack list";
+  const status = runSpackCommand(command);
+}
+
+/* Queries spack for package existance */
+private proc searchSpkgs(args: [?d] string) {
```

Comment:
This is vulnerable to shell injection.

```bash
mason external search "foo && rm Mason.toml"
```

 I'm not sure how bad that is in this case, but it's generally a good idea to avoid / prevent this.

Reply:
Long term, we want to use `spawn`. I'd put a `TODO` on the `spawnshell` case, and not worry about this too much right now.